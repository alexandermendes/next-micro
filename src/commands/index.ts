import { dev, DevCommand, DevArgs } from './dev';

export type Command = DevCommand;
export type CommandArgs = DevArgs;

export { dev };
