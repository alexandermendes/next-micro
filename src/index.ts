export { cli } from './cli';
export { NextMicroConfig } from './config';
