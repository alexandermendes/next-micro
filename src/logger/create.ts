import { sep } from 'path';
import { format } from 'util';
import chalk from 'chalk';
import figures from 'figures';

type ConsoleArg = string | Error;

type LogFunction = {
  (...args: ConsoleArg[]): void;
};

export type Logger = {
  [index: string]: LogFunction;
  error: LogFunction;
  warn: LogFunction;
  info: LogFunction;
  success: LogFunction;
  log: LogFunction;
  debug: LogFunction;
};

enum LogLevel {
  ERROR,
  WARN,
  INFO,
  SUCCESS,
  LOG,
  DEBUG,
}

export type LogLevelStrings = keyof typeof LogLevel;

export type LoggerOptions =
  | Record<string, never>
  | {
      logLevel?: string;
      icons?: boolean;
      colours?: boolean;
    };

const formatStack = (stack: string) => {
  const cwd = process.cwd() + sep;

  const lines = stack
    .split('\n')
    .splice(1)
    .map((line: string) => line.trim().replace('file://', '').replace(cwd, ''));

  return `  ${lines.join('\n  ')}`;
};

const formatArgs = (args: ConsoleArg[]) => {
  const _args = args.map((arg: ConsoleArg) => {
    if (arg instanceof Error && typeof arg.stack === 'string') {
      return arg.message + '\n' + formatStack(arg.stack);
    }

    return arg;
  });

  return format(..._args);
};

const writeStream = (level: number, msg: string) => {
  if (level < 1) {
    process.stderr.write(msg);

    return;
  }

  process.stdout.write(msg);
};

const getActiveLogLevel = (logOpts: LoggerOptions): LogLevel => {
  const activeLevelKey = String(logOpts.logLevel).toUpperCase();

  return LogLevel[<LogLevelStrings>activeLevelKey];
};

const createLogFunction =
  (
    logOpts: LoggerOptions,
    level: LogLevel,
    icon: string | null = null,
    color = chalk.white,
  ): LogFunction =>
  (...args: ConsoleArg[]) => {
    let message = formatArgs(args);

    if (icon && logOpts.icons) {
      message = `${figures(icon)} ${message}`;
    }

    if (logOpts.colours) {
      message = color(message);
    }

    if (level > getActiveLogLevel(logOpts)) {
      return;
    }

    writeStream(level, `${message}\n`);
  };

/**
 * A basic custom logger that prettifies log messages.
 */
export const createLogger = (logOpts: LoggerOptions = {}): Logger => {
  const defaultLogOpts = {
    logLevel: 'log',
    icons: true,
    colours: true,
  };

  const opts = Object.assign({}, defaultLogOpts, logOpts);

  return {
    error: createLogFunction(opts, LogLevel.ERROR, '✗', chalk.red),
    warn: createLogFunction(opts, LogLevel.WARN, '⚠', chalk.yellow),
    info: createLogFunction(opts, LogLevel.INFO, 'ℹ', chalk.cyan),
    success: createLogFunction(opts, LogLevel.SUCCESS, '✔', chalk.green),
    log: createLogFunction(opts, LogLevel.LOG),
    debug: createLogFunction(opts, LogLevel.DEBUG, '›'),
  };
};
