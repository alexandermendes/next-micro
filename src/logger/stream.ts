import { Writable } from 'stream';
import { LogFunction } from './types';

/**
 * Create a log stream that uses a custom log function.
 */
export const createLogStream = (log: LogFunction): Writable => {
  const writableStream = new Writable();

  writableStream._write = (chunk, encoding, next) => {
    log(chunk.toString().trim());
    next();
  };

  return writableStream;
};
