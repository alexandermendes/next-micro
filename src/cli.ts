import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { Command, dev } from './commands';
import { DevArgs } from './commands/dev';
import { logger } from './logger';

/**
 * Wrap async commands and ensure we exit the CLI with a non-zero exit code.
 */
const wrapAsync = (cmd: Command) => async (args: DevArgs) => {
  try {
    await cmd(args);
  } catch (err) {
    // eslint-disable-next-line no-console
    logger.error(err);

    process.exit(1);
  }
};

/**
 * Run the CLI.
 */
export const cli = (): void => {
  yargs(hideBin(process.argv))
    .command(
      'dev [names...]',
      'launch the local dev server',
      (yargs: yargs.Argv) => {
        return yargs.positional('names', {
          describe: 'a list of service names',
          default: [],
        });
      },
      wrapAsync(dev),
    )
    .option('env', {
      type: 'string',
      description: 'an environment name (see the env_* config setting)',
    })
    .strictCommands()
    .demandCommand(1).argv;
};
