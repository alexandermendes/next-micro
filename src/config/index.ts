import { cosmiconfig } from 'cosmiconfig';
import appRoot from 'app-root-path';

import { validate } from './validation';
import { logger } from '../logger';

export type ServiceConfig = {
  rootDir: string;
  name?: string;
  port?: number;
  routes?: string[];
  script?: string;
  scriptWaitTimeout?: number;
  watch?: boolean;
  ttl?: number;
  env?: Record<string, unknown>;
};

export type MicroproxyConfig = {
  port?: number;
  autoload?: boolean;
  autostart?: boolean;
  services?: ServiceConfig[];
};

export type ConcreteMicroproxyConfig = {
  port: number;
  autoload: boolean;
  autostart: boolean;
  services: ServiceConfig[];
};

const defaults = {
  autoload: true,
  port: 3000,
  autostart: false,
  services: [],
};

/**
 * Load the microproxy config.
 */
export const loadConfig = async (): Promise<ConcreteMicroproxyConfig> => {
  const dir = appRoot.path;
  const explorer = cosmiconfig('microproxy', { stopDir: dir });
  const { config } = (await explorer.search(dir)) || {};

  validate(config);

  return {
    ...defaults,
    ...config,
  };
};
