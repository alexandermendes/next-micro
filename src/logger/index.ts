import { createLogger } from './create';

export { createLogger };

export const logger = createLogger({
  logLevel: process.env.MICROPROXY_LOG_LEVEL,
});
