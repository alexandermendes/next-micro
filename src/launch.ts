import { loadConfig } from './config/load';
import { ProxyServer } from './proxy/server';
import { Router } from './router';
import { createServices } from './services/create';

/**
 * Launch the proxy server.
 */
export const launch = async (devMode = false): Promise<ProxyServer> => {
  const config = await loadConfig();
  const services = createServices(config.services);
  const router = new Router(services);
  const server = new ProxyServer(router, devMode);

  return server.launch(config.port);
};
