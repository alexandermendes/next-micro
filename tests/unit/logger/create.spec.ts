import { createLogger } from '../../../src/logger';

const originalStdOutWrite = process.stdout.write;
const originalStdErrWrite = process.stderr.write;

describe('Logger', () => {
  beforeAll(() => {
    process.stdout.write = jest.fn();
    process.stderr.write = jest.fn();
  });

  afterAll(() => {
    process.stdout.write = originalStdOutWrite;
    process.stderr.write = originalStdErrWrite;
  });

  it.each(['warn', 'info', 'success', 'log', 'debug'])(
    'logs %s to stdout',
    (type) => {
      const writeSpy = jest.spyOn(process.stdout, 'write');
      const logger = createLogger({ logLevel: 'DEBUG' });

      logger[type]('Hello');

      expect(writeSpy).toHaveBeenCalledTimes(1);
      expect(writeSpy.mock.calls[0][0]).toEqual(
        expect.stringContaining('Hello'),
      );
    },
  );

  it('logs errors to stderr', () => {
    const writeSpy = jest.spyOn(process.stderr, 'write');
    const logger = createLogger({ logLevel: 'DEBUG' });

    logger.error(new Error('bad thing'));

    expect(writeSpy).toHaveBeenCalledTimes(1);
    expect(String(writeSpy.mock.calls[0][0])).toEqual(
      expect.stringContaining('bad thing'),
    );
  });

  it('does not log messages below the log level', () => {
    const writeSpy = jest.spyOn(process.stdout, 'write');

    const logger = createLogger({ logLevel: 'ERROR' });
    logger.warn('Hello');

    expect(writeSpy).not.toHaveBeenCalled();
  });

  it('logs messages at the log level', () => {
    const writeSpy = jest.spyOn(process.stdout, 'write');

    const logger = createLogger({ logLevel: 'WARN' });
    logger.warn('Hello');

    expect(writeSpy).toHaveBeenCalled();
  });

  it('disables colours', () => {
    const writeSpy = jest.spyOn(process.stdout, 'write');

    const logger = createLogger({
      logLevel: 'LOG',
      colours: false,
    });

    logger.log('Hello');

    expect(writeSpy).toHaveBeenCalledWith('Hello\n');
  });

  it('creates a logger that with some reasonable defauts', () => {
    const writeSpy = jest.spyOn(process.stdout, 'write');
    const logger = createLogger();

    logger.log('log');
    logger.debug('debug');

    expect(writeSpy).toHaveBeenCalledTimes(1);
    expect(String(writeSpy.mock.calls[0][0])).toContain('log');
  });

  it('creates a logger with a tag', () => {
    const writeSpy = jest.spyOn(process.stdout, 'write');
    const logger = createLogger({ tag: 'service: my-service' });

    logger.log('Hello');

    expect(writeSpy).toHaveBeenCalledTimes(1);
    expect(writeSpy.mock.calls[0][0]).toEqual(
      expect.stringContaining('[service: my-service] Hello'),
    );
  });

  it('adds the tag to all stderr lines for an Error object', () => {
    const writeSpy = jest.spyOn(process.stderr, 'write');
    const logger = createLogger({
      tag: 'service: my-service',
      colours: false,
    });

    logger.error(new Error('bad thing'));

    const lines = String(writeSpy.mock.calls[0][0]).split('\n');
    const populatedLines = lines.filter((x) => x);

    expect(writeSpy).toHaveBeenCalledTimes(1);
    expect(populatedLines.length).toBeGreaterThan(1);
    expect(
      populatedLines.every((line) => line.startsWith('[service: my-service]')),
    ).toBe(true);
  });

  it('adds the tag to all stderr lines for a string with newlines', () => {
    const writeSpy = jest.spyOn(process.stderr, 'write');
    const logger = createLogger({
      tag: 'service: my-service',
      colours: false,
    });

    logger.error('I\nhave\nmultiple\nlines');

    const lines = String(writeSpy.mock.calls[0][0]).split('\n');
    const populatedLines = lines.filter((x) => x);

    expect(writeSpy).toHaveBeenCalledTimes(1);
    expect(populatedLines.length).toBeGreaterThan(1);
    expect(
      populatedLines.every((line) => line.startsWith('[service: my-service]')),
    ).toBe(true);
  });

  it('adds the tag to all stdout lines for a string with newlines', () => {
    const writeSpy = jest.spyOn(process.stdout, 'write');
    const logger = createLogger({
      tag: 'service: my-service',
      colours: false,
    });

    logger.log('I\nhave\nmultiple\nlines');

    const lines = String(writeSpy.mock.calls[0][0]).split('\n');
    const populatedLines = lines.filter((x) => x);

    expect(writeSpy).toHaveBeenCalledTimes(1);
    expect(populatedLines.length).toBeGreaterThan(1);
    expect(
      populatedLines.every((line) => line.startsWith('[service: my-service]')),
    ).toBe(true);
  });
});
