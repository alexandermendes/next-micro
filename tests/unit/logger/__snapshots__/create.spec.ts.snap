// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger logs debug to stdout 1`] = `
Array [
  "[37mâ€º Hello[39m
",
]
`;

exports[`Logger logs errors to stderr 1`] = `"[31mâœ— bad thing[39m"`;

exports[`Logger logs info to stdout 1`] = `
Array [
  "[36mâ„¹ Hello[39m
",
]
`;

exports[`Logger logs log to stdout 1`] = `
Array [
  "[37mHello[39m
",
]
`;

exports[`Logger logs success to stdout 1`] = `
Array [
  "[32mâœ” Hello[39m
",
]
`;

exports[`Logger logs warn to stdout 1`] = `
Array [
  "[33mâš  Hello[39m
",
]
`;
